File: app/src/main/java/com/scatl/uestcbbs/module/message/adapter/PrivateChatAdapter.java
package com.scatl.uestcbbs.module.message.adapter;

import android.content.Context;
import android.text.Html;
import android.view.View;
import android.widget.TextView;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.custom.postview.MyImageGetter;
import com.scatl.uestcbbs.entity.PrivateChatBean;
import com.scatl.uestcbbs.helper.glidehelper.GlideLoader4Common;
import com.scatl.uestcbbs.util.SharePrefUtil;
import com.scatl.uestcbbs.util.TimeUtil;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * author: sca_tl
 * description:
 * date: 2020/1/29 19:13
 */
public class PrivateChatAdapter extends BaseQuickAdapter<PrivateChatBean.BodyBean.PmListBean.MsgListBean, BaseViewHolder> {

    private String hisName, hisAvatar;
    private int hisUid;

    public PrivateChatAdapter(int layoutResId) {
        super(layoutResId);
    }

    public void setHisInfo(String hisName, String hisAvatar, int hisUid) {
        this.hisName = hisName;
        this.hisAvatar = hisAvatar;
        this.hisUid = hisUid;
    }

    public void insertMsg(Context context, String content, String type) {

        PrivateChatBean.BodyBean.PmListBean.MsgListBean msgListBean = new PrivateChatBean.BodyBean.PmListBean.MsgListBean();
        msgListBean.type = type;
        msgListBean.sender = SharePrefUtil.getUid(context);
        msgListBean.time = String.valueOf(System.currentTimeMillis());
        msgListBean.content = content;

        addData(msgListBean);
    }

    public void deleteMsg(int position) {
        getData().remove(position);
        notifyItemRemoved(position);
        notifyItemRangeChanged(position, getData().size());
    }

    @Override
    protected void convert(BaseViewHolder helper, PrivateChatBean.BodyBean.PmListBean.MsgListBean item) {

        int mine_id = SharePrefUtil.getUid(mContext);
        String mine_avatar = SharePrefUtil.getAvatar(mContext);

        if (item.sender == mine_id) {
            helper.addOnClickListener(R.id.item_private_chat_mine_img);
            helper.getView(R.id.item_private_chat_his_rl).setVisibility(View.GONE);
            helper.getView(R.id.item_private_chat_mine_rl).setVisibility(View.VISIBLE);

            GlideLoader4Common.simpleLoad(mContext, mine_avatar, helper.getView(R.id.item_private_chat_mine_icon));

            if (item.type.equals("text")) {
                helper.getView(R.id.item_private_chat_mine_content).setVisibility(View.VISIBLE);
                helper.getView(R.id.item_private_chat_mine_img).setVisibility(View.GONE);
//                helper.setText(R.id.item_private_chat_mine_content, item.content);
                setTextWithEmotion(helper.getView(R.id.item_private_chat_mine_content), item.content, false);

            }
            if (item.type.equals("image")) {
                helper.getView(R.id.item_private_chat_mine_content).setVisibility(View.GONE);
                helper.getView(R.id.item_private_chat_mine_img).setVisibility(View.VISIBLE);
                GlideLoader4Common.simpleLoad(mContext, item.content, helper.getView(R.id.item_private_chat_mine_img));
            }

            helper.setText(R.id.item_private_chat_mine_time, TimeUtil.formatTime(item.time, R.string.post_time1, mContext));

        } else {
            helper.addOnClickListener(R.id.item_private_chat_his_img);
            helper.getView(R.id.item_private_chat_mine_rl).setVisibility(View.GONE);
            helper.getView(R.id.item_private_chat_his_rl).setVisibility(View.VISIBLE);

            if (item.type.equals("text")) {
                helper.getView(R.id.item_private_chat_his_content).setVisibility(View.VISIBLE);
                helper.getView(R.id.item_private_chat_his_img).setVisibility(View.GONE);
//                helper.setText(R.id.item_private_chat_his_content, item.content);
                setTextWithEmotion(helper.getView(R.id.item_private_chat_his_content), item.content, false);
            }
            if (item.type.equals("image")) {
                helper.getView(R.id.item_private_chat_his_content).setVisibility(View.GONE);
                helper.getView(R.id.item_private_chat_his_img).setVisibility(View.VISIBLE);
                GlideLoader4Common.simpleLoad(mContext, item.content, helper.getView(R.id.item_private_chat_his_img));
            }

            GlideLoader4Common.simpleLoad(mContext, hisAvatar, helper.getView(R.id.item_private_chat_his_icon));

            helper.setText(R.id.item_private_chat_his_time,
                    TimeUtil.formatTime(item.time, R.string.post_time1, mContext));

        }
    }

    private void setTextWithEmotion(final TextView textView, String text, boolean append) {
        final Matcher matcher = Pattern.compile("(\\[mobcent_phiz=(.*?)])").matcher(text);

        if (matcher.find()) {
            do {
                text = text.replace(matcher.group(0)+"", "<img src = " + matcher.group(2) + ">");
            } while (matcher.find());
            text = text.replaceAll("\n", "<br>");
            if (append) {
                textView.append(Html.fromHtml(text, new MyImageGetter(mContext, textView), null));
            } else {
                textView.setText(Html.fromHtml(text, new MyImageGetter(mContext, textView), null));
            }

        } else {
            if (append) {
                textView.append(text);
            } else {
                textView.setText(text);
            }
        }

    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/message/adapter/PrivateChatAdapter.kt
package com.scatl.uestcbbs.module.message.adapter

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.material.shape.ShapeAppearanceModel
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.databinding.ItemPrivateChatBinding
import com.scatl.uestcbbs.entity.PrivateChatBean.BodyBean.PmListBean.MsgListBean
import com.scatl.uestcbbs.helper.PreloadAdapter
import com.scatl.uestcbbs.helper.ViewBindingHolder
import com.scatl.uestcbbs.util.Constant
import com.scatl.uestcbbs.util.SharePrefUtil
import com.scatl.uestcbbs.util.TimeUtil
import com.scatl.uestcbbs.util.load
import com.scatl.util.ColorUtil
import com.scatl.util.NumberUtil
import com.scatl.util.ScreenUtil

/**
 * Created by sca_tl at 2023/3/30 9:24
 */
class PrivateChatAdapter : PreloadAdapter<MsgListBean, ItemPrivateChatBinding>() {

    override fun getViewBinding(parent: ViewGroup): ItemPrivateChatBinding {
        return ItemPrivateChatBinding.inflate(LayoutInflater.from(context), parent, false)
    }

    fun insertMsg(context: Context, content: String?, type: String) {
        val msgBean = MsgListBean().apply {
            this.type = type
            this.sender = SharePrefUtil.getUid(context)
            this.time = System.currentTimeMillis().toString()
            this.content = content
        }
        add(msgBean)
    }

    override fun onBindViewHolder(holder: ViewBindingHolder<ItemPrivateChatBinding>, position: Int, item: MsgListBean?) {
        super.onBindViewHolder(holder, position, item)
        if (item == null) {
            return
        }
        if (item.sender == SharePrefUtil.getUid(context)) {
            holder.binding.hisContentLayout.visibility = View.GONE
            holder.binding.mineContentLayout.visibility = View.VISIBLE
            setMineContent(holder, item)
        } else {
            holder.binding.hisContentLayout.visibility = View.VISIBLE
            holder.binding.mineContentLayout.visibility = View.GONE
            setHisContent(holder, item)
        }
    }

    private fun setMineContent(holder: ViewBindingHolder<ItemPrivateChatBinding>, item: MsgListBean) {
        holder.binding.mineIcon.load(Constant.USER_AVATAR_URL.plus(item.sender))
        holder.binding.mineTime.text = TimeUtil.getFormatDate(NumberUtil.parseLong(item.time), "yyyy-MM-dd HH:mm")

        val cornerR = ScreenUtil.dip2pxF(context, 15f)
        holder.binding.mineContentBg.background = GradientDrawable().apply {
            setColor(ColorUtil.getAttrColor(context, R.attr.colorPrimaryContainer))
            cornerRadii = floatArrayOf(cornerR, cornerR, 0f, 0f, cornerR, cornerR, cornerR, cornerR)
        }

        if (item.type == "text") {
            holder.binding.mineTextContent.visibility = View.VISIBLE
            holder.binding.mineImgContent.visibility = View.GONE
            holder.binding.mineTextContent.setText(item.content)
            holder.binding.mineContentBg.backgroundTintList = ColorStateList.valueOf(ColorUtil.getAttrColor(context, R.attr.colorPrimaryContainer))
        } else if (item.type == "image") {
            holder.binding.mineTextContent.visibility = View.GONE
            holder.binding.mineImgContent.visibility = View.VISIBLE
            holder.binding.mineContentBg.backgroundTintList = ColorStateList.valueOf(ColorUtil.getAttrColor(context, R.attr.colorSurface))
            holder.binding.mineImgContent.load(item.content)
            holder.binding.mineImgContent.shapeAppearanceModel = ShapeAppearanceModel.builder().apply {
                setTopLeftCornerSize(cornerR)
                setTopRightCornerSize(0f)
                setBottomLeftCornerSize(0f)
                setBottomRightCornerSize(0f)
            }.build()
        }
    }

    private fun setHisContent(holder: ViewBindingHolder<ItemPrivateChatBinding>, item: MsgListBean) {
        holder.binding.hisIcon.load(Constant.USER_AVATAR_URL.plus(item.sender))
        holder.binding.hisTime.text = TimeUtil.getFormatDate(NumberUtil.parseLong(item.time), "yyyy-MM-dd HH:mm")

        val cornerR = ScreenUtil.dip2pxF(context, 15f)
        holder.binding.hisContentBg.background = GradientDrawable().apply {
            setColor(ColorUtil.getAttrColor(context, R.attr.colorSurface))
            cornerRadii = floatArrayOf(0f, 0f, cornerR, cornerR, cornerR, cornerR, cornerR, cornerR)
        }

        if (item.type == "text") {
            holder.binding.hisTextContent.visibility = View.VISIBLE
            holder.binding.hisImgContent.visibility = View.GONE
            holder.binding.hisTextContent.setText(item.content)
        } else if (item.type == "image") {
            holder.binding.hisTextContent.visibility = View.GONE
            holder.binding.hisImgContent.visibility = View.VISIBLE
            holder.binding.hisImgContent.load(item.content)
            holder.binding.hisImgContent.shapeAppearanceModel = ShapeAppearanceModel.builder().apply {
                setTopLeftCornerSize(0f)
                setTopRightCornerSize(cornerR)
                setBottomLeftCornerSize(0f)
                setBottomRightCornerSize(0f)
            }.build()
        }
    }
}